src/main/java/io/quarkus/gizmo/StringSwitchImpl.java

```diff
@@ -0,0 +1,120 @@
+package io.quarkus.gizmo;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import org.objectweb.asm.Label;
+import org.objectweb.asm.MethodVisitor;
+
+class StringSwitchImpl extends BytecodeCreatorImpl implements StringSwitch {
+
+    private final Map<Integer, List<Entry<String, Consumer<BytecodeCreator>>>> hashToCaseBlocks;
+
+    public StringSwitchImpl(ResultHandle value, Consumer<BytecodeCreator> defaultBlockConsumer, BytecodeCreatorImpl enclosing) {
+        super(enclosing);
+        this.hashToCaseBlocks = new LinkedHashMap<>();
+        ResultHandle strHash = invokeVirtualMethod(MethodDescriptor.ofMethod(Object.class, "hashCode", int.class), value);
+
+        Set<ResultHandle> inputHandles = new HashSet<>();
+        inputHandles.add(value);
+        inputHandles.add(strHash);
+
+        operations.add(new Operation() {
+
+            @Override
+            void writeBytecode(MethodVisitor methodVisitor) {
+                Map<Integer, Label> hashToLabel = new HashMap<>();
+                List<BytecodeCreatorImpl> lookupBlocks = new ArrayList<>();
+                Map<String, BytecodeCreatorImpl> caseBlocks = new LinkedHashMap<>();
+                BytecodeCreatorImpl defaultBlock = new BytecodeCreatorImpl(StringSwitchImpl.this);
+                if (defaultBlockConsumer != null) {
+                    defaultBlockConsumer.accept(defaultBlock);
+                }
+
+                // Initialize the case blocks and lookup blocks
+                for (Entry<Integer, List<Entry<String, Consumer<BytecodeCreator>>>> hashEntry : hashToCaseBlocks.entrySet()) {
+                    BytecodeCreatorImpl lookupBlock = new BytecodeCreatorImpl(StringSwitchImpl.this);
+                    for (Entry<String, Consumer<BytecodeCreator>> caseEntry : hashEntry.getValue()) {
+                        BytecodeCreatorImpl caseBlock = new BytecodeCreatorImpl(StringSwitchImpl.this);
+                        caseEntry.getValue().accept(caseBlock);
+                        caseBlock.findActiveResultHandles(inputHandles);
+                        caseBlocks.put(caseEntry.getKey(), caseBlock);
+                        BytecodeCreator isEqual = lookupBlock
+                                .ifTrue(Gizmo.equals(lookupBlock, lookupBlock.load(caseEntry.getKey()), value)).trueBranch();
+                        isEqual.jumpTo(caseBlock);
```

Comment:
Wouldn't it be better to create the case block as the child of the `if`? (I don't remember if that is possible...)

Reply:
I'm not sure. TBH I've tried to mimic the behavior of `javac` because I would expect that it's a reasonable approach...